source/simulator/assembly.cc

```diff
@@ -973,13 +367,17 @@ namespace aspect
       }
 
     // allow other assemblers to add themselves or modify the existing ones by firing the signal
-    this->signals.set_assemblers(*this, *assemblers, assembler_objects);
+    this->signals.set_assemblers(*this, *assemblers);
 
     // ensure that all assembler objects have access to the SimulatorAccess
     // base class
-    for (unsigned int i=0; i<assembler_objects.size(); ++i)
-      if (SimulatorAccess<dim> *p = dynamic_cast<SimulatorAccess<dim>*>(assembler_objects[i].get()))
-        p->initialize_simulator(*this);
+    initialize_simulator(*this,assemblers->stokes_preconditioner);
+    initialize_simulator(*this,assemblers->stokes_system);
+    initialize_simulator(*this,assemblers->stokes_system_on_boundary_face);
+    initialize_simulator(*this,assemblers->advection_system);
+    initialize_simulator(*this,assemblers->advection_system_on_boundary_face);
+    initialize_simulator(*this,assemblers->advection_system_on_interior_face);
+    initialize_simulator(*this,assemblers->advection_system_residual);
```

Comment:
Hm, this is not really nice. Isn't there a way to do this for all of them at once?

Reply:
It turns out there is a way, but for that I need to implement a constructor for all assemblers (this constructor would take the `this` pointer to the simulator and call the constructor of the `SimulatorAccess` class that immediately does what `initialize_simulator` does). But this needs more lines than I have here. Still if you would prefer the constructor I can implement that (it might look cleaner, but require more lines).

Reply:
Hm. What kind of error message do you get if you forget to call this line for your assembler? I am just thinking, if you make a new type of assembler and forget to implement this line, how easy is it to find out what the problem is? If that is obvious, I would be okay with leaving it like it is.